### PR TITLE
Move doc-comment highlight injection from AST to HIR

### LINKED ISSUE #5016: Doctest syntax highlighting does not work with `#[doc]` attributes
`cfg` and friends paired with `doc = "..."` should trigger syntax highlight if applicable.

```rust
/// # Example
///
/// ```rust
/// let _ = example(&[1, 2, 3]);
/// ```
///
/// # Another example
///
#[cfg_attr(feature = "alloc", doc = "```rust")]
#[cfg_attr(not(feature = "alloc"), doc = "```ignore")]
/// let _ = example(&alloc::vec![1, 2, 3]);
/// ```
pub fn example<T>(t: &T) -> &[i32]
where
    T: AsRef<[i32]>
{
  t.as_ref()
}
```

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #5016

### CODE REVIEW

**Comment by lnicola** [comment]
changelog internal

**Comment by Veykril** [comment]
image
Gotta say I didn't expect injection to just work in attributes after fixing up some text ranges

File: crates/ide/src/syntax_highlighting/inject.rs [hunk 1]
```diff
@@ -101,12 +163,33 @@ pub(super) fn doc_comment(hl: &mut Highlights, node: &SyntaxNode) {
     // Replace the original, line-spanning comment ranges by new, only comment-prefix
     // spanning comment ranges.
     let mut new_comments = Vec::new();
-    for comment in doc_comments {
-        match comment.text().find(RUSTDOC_FENCE) {
+    let mut string;
+    for attr in attributes.by_key("doc").attrs() {
+        let src = attr.to_src(&owner);
```
Comment (Veykril) [inline crates/ide/src/syntax_highlighting/inject.rs:168]: This `to_src` call actually has O(n^2) runtime as it recollects all doc items every call then takes the nth item from there, should probably look into a better mapping

Comment (jonas-schievink) [reply crates/ide/src/syntax_highlighting/inject.rs:168]: Yeah, I'm pretty unsure what would make a good API for attributes. Let me know if you have ideas.

File: crates/ide/src/syntax_highlighting/inject.rs [hunk 2]
```diff
@@ -156,3 +236,27 @@ pub(super) fn doc_comment(hl: &mut Highlights, node: &SyntaxNode) {
...
+fn find_doc_string_in_attr(attr: &hir::Attr, it: &ast::Attr) -> Option<ast::String> {
+    match it.literal() {
+        // #[doc = lit]
+        Some(lit) => match lit.kind() {
+            ast::LiteralKind::String(it) => Some(it),
+            _ => None,
+        },
+        // #[cfg_attr(..., doc = "", ...)]
+        None => {
+            // We gotta hunt the string token manually here
```
Comment (jonas-schievink) [inline crates/ide/src/syntax_highlighting/inject.rs:249]: This is also something I'd love to fix with a better attributes API

File: crates/ide/src/syntax_highlighting/inject.rs (hunk 1 above)
Comment (matklad) [reply crates/ide/src/syntax_highlighting/inject.rs:168]: Yeah, most of our `.source` calls are quadratic. The fundamental thing here is that we can't ask for sources of a *single* thing. Rather, we can ask the parent "what are the sources of all your children?", and there needs to be some caching layer there. `impl HasSource for Variant` is also quadratic, for example, and that's not really right. 

Ideally, this caching should be done transparently by salsa, but this is the bit where we need gc unfortunately.

File: crates/hir/src/semantics/source_to_def.rs [hunk 3]
```diff
@@ -71,6 +71,12 @@ impl SourceToDefCtx<'_, '_> {
         Some(def_map.module_id(child_id))
     }
 
+    pub(super) fn source_file_to_def(&mut self, src: InFile<ast::SourceFile>) -> Option<ModuleId> {
+        let _p = profile::span("source_file_to_def");
+        let file_id = src.file_id.original_file(self.db.upcast());
+        self.file_to_def(file_id).get(0).copied()
+    }
```
[the remark this answers is not in the document]
Comment (Veykril) [reply crates/hir/src/semantics/source_to_def.rs:78]: I'm not sure how I should be doing that then, the caller here is still Semantics itself in the generic `to_def` call, and prior to the `to_def` call I don't have the InFile information as thats something semantics looks for itself

[the remark this answers is not in the document]
Comment (matklad) [reply crates/hir/src/semantics/source_to_def.rs:78]: ah, sorry, I missed that this is `pub(super)`, LGTM then!

**Comment by Veykril** [comment]
bors r=matklad,jonas-schievink

**Pull request merged**